python/lsst/source/injection/inject_engine.py

```diff
@@ -0,0 +1,462 @@
+        A fully specified and transformed GalSim object.
+    """
+    if logger:
+        source_types = Counter(injection_catalog["source_type"])  # type: ignore
+        grammar0 = "source" if len(injection_catalog) == 1 else "sources"
+        grammar1 = "type" if len(source_types) == 1 else "types"
+        logger.info(
+            "Generating %d injection %s consisting of %d unique %s: %s.",
+            len(injection_catalog),
+            grammar0,
+            len(source_types),
+            grammar1,
+            ", ".join(f"{k}({v})" for k, v in source_types.items()),
+        )
+    for source_data_full in injection_catalog:
+        items = dict(source_data_full).items()  # type: ignore
+        source_data = {k: v for (k, v) in items if v is not ma.masked}
+        try:
+            sky_coords = SpherePoint(float(source_data["ra"]), float(source_data["dec"]), degrees)
+        except KeyError:
+            sky_coords = wcs.pixelToSky(float(source_data["x"]), float(source_data["y"]))
+        try:
+            pixel_coords = Point2D(source_data["x"], source_data["y"])
+        except KeyError:
+            pixel_coords = wcs.skyToPixel(sky_coords)
+        try:
+            inst_flux = photo_calib.magnitudeToInstFlux(source_data["mag"], pixel_coords)
+        except LogicError:
+            continue
+        try:
+            draw_size = int(source_data["draw_size"])
+        except KeyError:
+            draw_size = 0
+
+        if source_data["source_type"] == "Stamp":
+            stamp_file = stamp_prefix + source_data["stamp"]
+            object = make_galsim_stamp(stamp_file, fits_alignment, wcs, sky_coords, inst_flux)
+        elif source_data["source_type"] == "Trail":
+            object = make_galsim_trail(source_data, wcs, sky_coords, inst_flux)
+        else:
+            object = make_galsim_object(source_data, source_data["source_type"], inst_flux)
+
+        yield sky_coords, pixel_coords, draw_size, object
+
+
+def make_galsim_object(
+    source_data: dict[str, Any],
+    source_type: str,
+    inst_flux: float,
+) -> galsim.transform.Transformation:
+    """Make a generic GalSim object from a collection of source data.
+
+    Parameters
+    ----------
+    source_data : `dict` [`str`, `Any`]
+        Dictionary of source data.
+    source_type : `str`
+        Type of the source, corresponding to a GalSim class.
+    inst_flux : `float`
+        Instrumental flux of the source.
+
+    Returns
+    -------
+    object : `galsim.transform.Transformation`
+        A fully specified and transformed GalSim object.
+    """
+    # Populate the non-shear and non-flux parameters.
+    object_class = getattr(galsim, source_type)
+    object_params = set([x for x in dir(object_class) if not x.startswith("_")]) & set(source_data.keys())
+    object_data = dict()
+    for object_param in object_params:
+        object_data.update({object_param: source_data[object_param]})
+    object = object_class(**object_data)
```

Comment:
Fine for now, but I think this really depends on internal implementation details of GalSim to work (and I'm not sure if it always works or just usually works for all GSObjects in fact).  I could easily imagine a GSObject taking a kwargs with name `name` and using that to set an attr `.name2`.  I think the right way to approach this is actually to use something like https://galsim-developers.github.io/GalSim/_build/html/config_process.html#galsim.config.BuildGSObject.  Please make a issue to investigate.

Reply:
:point_right: [DM-39725](https://jira.lsstcorp.org/browse/DM-39725)

Reply:
I think this should be close to what we want for determining arguments from a catalog row:

```python
def GetKwargs(row, cls):
    """Assemble dictionary of keyword arguments to use when constructing a 
    GSObject.
    """
    req = getattr(cls, '_req_params', {})
    opt = getattr(cls, '_opt_params', {})
    single = getattr(cls, '_single_params', {})

    params = {}

    # Check required args
    for key in req:
        if key not in row:
            raise ValueError(
                f"Required parameter {key} not found in input catalog"
            )
        params[key] = row[key]
    
    # Optional args
    for key in opt:
        if key in row:
            params[key] = row[key]
    
    # Single args
    for s in single:
        count = 0
        for key in s:
            if key in row:
                count += 1
                if count > 1:
                    raise ValueError(
                        f"Only one of {s.keys()} allowed for type {cls}"
                    )
                params[key] = row[key]
    if count == 0:
        raise ValueError(
            f"One of the args {s.keys()} is required for type {cls}"
        )
    return params
```

Reply:
Note that most `GSObject`s will accept `flux` as an argument, so we might want to explicitly strip that out of the output of `GetKwargs` above and just apply `.withFlux()` after constructing the object.  Fortunately, I don't think it's wrong to set the flux in both places, just redundant.

Reply:
Two new free functions have now been implemented in `inject_engine`: `get_object_data` and `get_shear_data`. The former (`get_object_data`) is essentially what you have above, bar a few variable name changes and small implementation details/complete doc-string/etc.

I've also added `get_shear_data` which does the same thing but for the Shear kwargs. To achieve this, as we discussed, I've just hard-coded the possible list of allowed shear attributes into the function.